tsdb/shard.go

```diff
@@ -356,10 +361,24 @@ func (s *Shard) Path() string { return s.path }
 
 // Open initializes and opens the shard's store.
 func (s *Shard) Open(ctx context.Context) error {
-	if err := func() error {
-		s.mu.Lock()
-		defer s.mu.Unlock()
+	s.mu.Lock()
+	closeWaitNeeded, err := s.openNoLock(ctx)
+	s.mu.Unlock()
+	if closeWaitNeeded {
+		werr := s.closeWait()
+		// We want the first error we get returned to the caller
+		if err == nil {
+			err = werr
+		}
+	}
+	return err
+}
 
+// openNoLock performs work of Open. Must hold s.mu before calling. The first return
+// value is true if the caller should call closeWait after unlocking s.mu in order
+// to clean up a failed open operation.
+func (s *Shard) openNoLock(ctx context.Context) (bool, error) {
```

Comment:
Is the bool return necessary? I only spot two `return` statements, with error and true and without error and false.

Reply:
I agree it's kind of clunky, but it's a future safety thing. We want to make it clear to callers of `openNoLock` that they will need to handle calling `closeWait`, but we also want to make sure that callers don't call `closeWait` unless `closeNoLock` was closed first. Currently all error return cases indicate that the caller should call `openNoLock`, but that may not be true in the future.

As the code stands right now, `openNoLock` could be eliminated, as the only place it is called from is `Open`. This made more sense when I originally implemented it, but the changes I had to make to handle the closing the `metricsUpdater` cause everything to call `Open` again.